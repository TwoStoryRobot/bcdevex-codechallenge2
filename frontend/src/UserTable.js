import React from 'react'
import PropTypes from 'prop-types'
import Avatar from 'react-avatar'
import { withStyles } from '@material-ui/core/styles'
import Paper from '@material-ui/core/Paper'
import Table from '@material-ui/core/Table'
import TableBody from '@material-ui/core/TableBody'
import TableCell from '@material-ui/core/TableCell'
import TableHead from '@material-ui/core/TableHead'
import TableRow from '@material-ui/core/TableRow'
import UserSearchIcon from 'mdi-react/UserSearchIcon'

const StyledTableCell = withStyles(theme => ({
  head: {
    fontSize: '1.25rem',
    color: theme.palette.text.hint,
    fontWeight: 400
  },
  body: {
    fontSize: '1.125rem',
    fontWeight: 300
  }
}))(TableCell)

const Message = withStyles(theme => ({
  root: {
    textAlign: 'center',
    margin: theme.spacing.unit * 3,
    color: theme.palette.text.secondary
  }
}))(({ classes, children, ...rest }) => (
  <div {...rest} className={classes.root}>
    {children}
  </div>
))

const styles = theme => ({
  root: {
    width: '100%',
    marginTop: theme.spacing.unit * 3,
    overflowX: 'auto'
  },
  table: {
    minWidth: 700
  },
  row: {
    '&:nth-of-type(odd)': {
      backgroundColor: theme.palette.background.default
    }
  }
})

/**
 *  A table to list all registered users
 */
const UserTable = ({ users, classes }) => (
  <Paper className={classes.root}>
    <Table className={classes.table}>
      <TableHead>
        <TableRow>
          <StyledTableCell />
          <StyledTableCell>First</StyledTableCell>
          <StyledTableCell>Last</StyledTableCell>
          <StyledTableCell>Email</StyledTableCell>
          <StyledTableCell>Registration</StyledTableCell>
        </TableRow>
      </TableHead>
      <TableBody>
        {users.map(user => (
          <TableRow key={user.userId} className={classes.row}>
            <StyledTableCell component="th" scope="row">
              <Avatar
                key={user.userId}
                src={user.imageUrl}
                name={`${user.firstName} ${user.lastName}`}
                size={34}
                round={true}
              />
            </StyledTableCell>
            <StyledTableCell>{user.firstName}</StyledTableCell>
            <StyledTableCell>{user.lastName}</StyledTableCell>
            <StyledTableCell>{user.emailAddress}</StyledTableCell>
            {/* TODO This needs to come from the user prop */}
            {/* Note: We are manually overriding this individual column font size */}
            <StyledTableCell style={{ fontSize: '0.875rem' }}>
              2018 Sept 04
            </StyledTableCell>
          </TableRow>
        ))}
      </TableBody>
    </Table>
    {users.length === 0 && (
      <Message>
        <UserSearchIcon size={64} />
        <p>There are currently no registered users.</p>
      </Message>
    )}
  </Paper>
)

UserTable.propTypes = {
  /** classes provided by the withStyles HOC */
  classes: PropTypes.object.isRequired,
  /** user object */
  users: PropTypes.arrayOf(
    PropTypes.shape({
      /** id of user */
      userId: PropTypes.string.isRequired,
      /** first name of user */
      firstName: PropTypes.string.isRequired,
      /** last name of user */
      lastName: PropTypes.string,
      /** email address of user (used to send email) */
      emailAddress: PropTypes.emailAddress,
      /** url of user's Avatar */
      imageUrl: PropTypes.string
    })
  ).isRequired
}

export default withStyles(styles)(UserTable)
