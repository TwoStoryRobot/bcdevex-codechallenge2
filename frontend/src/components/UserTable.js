import React from 'react'
import PropTypes from 'prop-types'
import Avatar from 'react-avatar'
import styled from 'styled-components'
import { withStyles } from '@material-ui/core/styles'
import Paper from '@material-ui/core/Paper'
import Table from '@material-ui/core/Table'
import TableBody from '@material-ui/core/TableBody'
import TableCell from '@material-ui/core/TableCell'
import TableHead from '@material-ui/core/TableHead'
import TableRow from '@material-ui/core/TableRow'
import Typography from '@material-ui/core/Typography'
import CircularProgress from '@material-ui/core/CircularProgress'
import IconButton from '@material-ui/core/IconButton'
import Button from '@material-ui/core/Button'
import UserSearchIcon from 'mdi-react/UserSearchIcon'
import DeleteIcon from 'mdi-react/DeleteIcon'
import PencilIcon from 'mdi-react/PencilIcon'
import EnvelopeIcon from 'mdi-react/EnvelopeIcon'
import SortDescendingIcon from 'mdi-react/SortAscendingIcon'
import SortAscendingIcon from 'mdi-react/SortDescendingIcon'

const StyledButtonsCell = withStyles(theme => ({
  root: {
    display: 'flex'
  }
}))(({ classes, children, ...rest }) => (
  <div {...rest} className={classes.root}>
    {children}
  </div>
))

const StyledTableCell = withStyles(theme => ({
  root: {
    padding: '4px 16px 4px 16px'
  },
  head: {
    fontSize: '1.25rem',
    color: theme.palette.text.hint,
    fontWeight: 400
  },
  body: {
    fontSize: '1.125rem',
    fontWeight: 300
  }
}))(TableCell)

const Message = withStyles(theme => ({
  root: {
    textAlign: 'center',
    margin: theme.spacing.unit * 3,
    color: theme.palette.text.secondary
  }
}))(({ classes, children, ...rest }) => (
  <div {...rest} className={classes.root}>
    {children}
  </div>
))

const styles = theme => ({
  root: {
    width: '100%',
    marginTop: theme.spacing.unit * 3,
    overflowX: 'auto'
  },
  table: {
    minWidth: 700
  },
  row: {
    '&:nth-of-type(odd)': {
      backgroundColor: theme.palette.background.default
    }
  },
  admin: {
    backgroundColor: '#f0c243'
  }
})

const AscIcon = styled(SortAscendingIcon)`
  margin-left: 8px;
`

const DescIcon = styled(SortDescendingIcon)`
  margin-left: 8px;
`

const EmptyIcon = styled.span`
  width: 32px;
  height: 24px;
`

/**
 *  A table to list all registered users
 */
class UserTable extends React.Component {
  state = {
    sortField: 'lastName',
    sortDirection: 'asc'
  }

  toggleSort = sortField => {
    if (this.state.sortField === sortField)
      this.setState({
        sortDirection: this.state.sortDirection === 'asc' ? 'desc' : 'asc'
      })
    else this.setState({ sortField, sortDirection: 'asc' })
  }

  getSortedUsers() {
    //clone a new array for users
    let users = [...this.props.users]
    if (this.state.sortDirection === 'asc')
      users.sort(this.genSortAscendingByField(this.state.sortField))
    else users.sort(this.genSortDescendingByField(this.state.sortField))

    return users
  }

  genSortAscendingByField(field) {
    return (a, b) => {
      const valA = a[field].toUpperCase()
      const valB = b[field].toUpperCase()
      let order = 0
      if (valA < valB) order = -1
      if (valB > valA) order = 1
      return order
    }
  }

  genSortDescendingByField(field) {
    return (a, b) => {
      const valA = a[field].toUpperCase()
      const valB = b[field].toUpperCase()
      let order = 0
      if (valA > valB) order = -1
      if (valB < valA) order = 1
      return order
    }
  }

  render() {
    const {
      users,
      classes,
      isLoading,
      isAdmin,
      handleEditClick,
      handleDeleteClick,
      handleSendEmail
    } = this.props

    const { sortField } = this.state
    const sortIcon =
      this.state.sortDirection === 'asc' ? <AscIcon /> : <DescIcon />

    return (
      <Paper className={classes.root}>
        <Table className={classes.table}>
          <TableHead>
            <TableRow>
              <StyledTableCell />
              <StyledTableCell>
                <Button onClick={() => this.toggleSort('firstName')}>
                  First
                  {sortField === 'firstName' ? sortIcon : <EmptyIcon />}
                </Button>
              </StyledTableCell>
              <StyledTableCell>
                <Button onClick={() => this.toggleSort('lastName')}>
                  Last
                  {sortField === 'lastName' ? sortIcon : <EmptyIcon />}
                </Button>
              </StyledTableCell>
              <StyledTableCell>
                <Button onClick={() => this.toggleSort('emailAddress')}>
                  Email
                  {sortField === 'emailAddress' ? sortIcon : <EmptyIcon />}
                </Button>
              </StyledTableCell>
              <StyledTableCell>
                <Typography variant="button">Registration</Typography>
              </StyledTableCell>
              {isAdmin && <StyledTableCell />}
            </TableRow>
          </TableHead>
          <TableBody>
            {this.getSortedUsers().map(user => (
              <TableRow key={user.userId} className={classes.row}>
                <StyledTableCell component="th" scope="row">
                  <Avatar
                    key={user.userId}
                    src={user.imageUrl}
                    name={`${user.firstName} ${user.lastName}`}
                    size={34}
                    round={true}
                  />
                </StyledTableCell>
                <StyledTableCell>{user.firstName}</StyledTableCell>
                <StyledTableCell>{user.lastName}</StyledTableCell>
                <StyledTableCell>{user.emailAddress}</StyledTableCell>
                {/* TODO This needs to come from the user prop */}
                {/* Note: We are manually overriding this individual column font size */}
                <StyledTableCell style={{ fontSize: '0.875rem' }}>
                  2018 Sept 04
                </StyledTableCell>
                {isAdmin && (
                  <StyledTableCell>
                    <StyledButtonsCell>
                      <IconButton
                        color="primary"
                        onClick={handleEditClick}
                        data-testid="edit">
                        <PencilIcon />
                      </IconButton>
                      <IconButton
                        color="primary"
                        onClick={handleDeleteClick}
                        data-testid="delete">
                        <DeleteIcon />
                      </IconButton>
                      <IconButton
                        color="primary"
                        onClick={handleSendEmail}
                        data-testid="email">
                        <EnvelopeIcon />
                      </IconButton>
                    </StyledButtonsCell>
                  </StyledTableCell>
                )}
              </TableRow>
            ))}
          </TableBody>
        </Table>
        {users.length === 0 &&
          !isLoading && (
            <Message>
              <UserSearchIcon size={64} />
              <p>There are currently no registered users.</p>
            </Message>
          )}
        {isLoading && (
          <Message>
            <CircularProgress size={64} />
            <p>Loading...</p>
          </Message>
        )}
      </Paper>
    )
  }
}

UserTable.propTypes = {
  /** classes provided by the withStyles HOC */
  classes: PropTypes.object.isRequired,
  /** user object */
  users: PropTypes.arrayOf(
    PropTypes.shape({
      /** id of user */
      userId: PropTypes.string.isRequired,
      /** first name of user */
      firstName: PropTypes.string.isRequired,
      /** last name of user */
      lastName: PropTypes.string,
      /** email address of user (used to send email) */
      emailAddress: PropTypes.emailAddress,
      /** url of user's Avatar */
      imageUrl: PropTypes.string,
      /** admin flag */
      isAdmin: PropTypes.bool
    })
  ).isRequired,
  isLoading: PropTypes.bool,
  isAdmin: PropTypes.bool.isRequired,
  handleEditClick: PropTypes.func.isRequired,
  handleDeleteClick: PropTypes.func.isRequired,
  handleSendEmail: PropTypes.func.isRequired
}

UserTable.defaultProps = {
  isLoading: false
}

export default withStyles(styles)(UserTable)
