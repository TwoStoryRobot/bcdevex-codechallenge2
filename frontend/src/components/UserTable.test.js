import React from 'react'
import { render, cleanup } from 'react-testing-library'
import 'jest-dom/extend-expect'

import UserTable from './UserTable'

afterEach(cleanup)

describe('UserTable', () => {
  const defaultUsers = [
    {
      userId: 'testUser123',
      firstName: 'Bob',
      lastName: 'Smith',
      emailAddress: 'bob.smith@gmail.com',
      imageUrl: 'https://d3iw72m71ie81c.cloudfront.net/male-52.jpg'
    },
    {
      userId: 'testUser234',
      firstName: 'Jane',
      lastName: 'Doe',
      emailAddress: 'jane.doe@gmail.com',
      imageUrl: 'https://randomuser.me/api/portraits/women/21.jpg'
    },

    {
      userId: 'testUser345',
      firstName: 'James',
      lastName: 'Bond',
      emailAddress: 'james.bond@gmail.com',
      imageUrl: 'https://twostoryrobot.com/not-a-valid-url'
    }
  ]

  it('should render with required props', async () => {
    const { container } = render(
      <UserTable
        users={defaultUsers}
        handleEditClick={() => {}}
        handleDeleteClick={() => {}}
        handleSendEmail={() => {}}
      />
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should display message when there are no registered users', async () => {
    const { container } = render(
      <UserTable
        users={[]}
        handleEditClick={() => {}}
        handleDeleteClick={() => {}}
        handleSendEmail={() => {}}
      />
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should display loading spinner when isLoaded is true', async () => {
    const { container } = render(
      <UserTable
        users={[]}
        isLoading={true}
        handleEditClick={() => {}}
        handleDeleteClick={() => {}}
        handleSendEmail={() => {}}
      />
    )

    expect(container.firstChild).toMatchSnapshot()
  })
})
