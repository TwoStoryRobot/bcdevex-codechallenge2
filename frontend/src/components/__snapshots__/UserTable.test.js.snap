// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserTable should display loading spinner when isLoaded is true 1`] = `
<div
  class="MuiPaper-root-0101 MuiPaper-elevation2-0105 MuiPaper-rounded-0102 UserTable-root-098"
>
  <table
    class="MuiTable-root-0128 UserTable-table-099"
  >
    <thead
      class="MuiTableHead-root-0129"
    >
      <tr
        class="MuiTableRow-root-0130 MuiTableRow-head-0133"
      >
        <th
          class="MuiTableCell-root-0137 MuiTableCell-head-0138 WithStyles-TableCell--head-0135"
          scope="col"
        />
        <th
          class="MuiTableCell-root-0137 MuiTableCell-head-0138 WithStyles-TableCell--head-0135"
          scope="col"
        >
          First
        </th>
        <th
          class="MuiTableCell-root-0137 MuiTableCell-head-0138 WithStyles-TableCell--head-0135"
          scope="col"
        >
          Last
        </th>
        <th
          class="MuiTableCell-root-0137 MuiTableCell-head-0138 WithStyles-TableCell--head-0135"
          scope="col"
        >
          Email
        </th>
        <th
          class="MuiTableCell-root-0137 MuiTableCell-head-0138 WithStyles-TableCell--head-0135"
          scope="col"
        >
          Registration
        </th>
      </tr>
    </thead>
    <tbody
      class="MuiTableBody-root-0145"
    />
  </table>
  <div
    class="Component-root-0146"
  >
    <div
      class="MuiCircularProgress-root-0147 MuiCircularProgress-colorPrimary-0150 MuiCircularProgress-indeterminate-0149"
      style="width: 64px; height: 64px;"
      role="progressbar"
    >
      <svg
        class="MuiCircularProgress-svg-0152"
        viewBox="22 22 44 44"
      >
        <circle
          class="MuiCircularProgress-circle-0153 MuiCircularProgress-circleIndeterminate-0155"
          cx="44"
          cy="44"
          r="20.2"
          fill="none"
          stroke-width="3.6"
        />
      </svg>
    </div>
    <p>
      Loading...
    </p>
  </div>
</div>
`;

exports[`UserTable should display message when there are no registered users 1`] = `
<div
  class="MuiPaper-root-052 MuiPaper-elevation2-056 MuiPaper-rounded-053 UserTable-root-049"
>
  <table
    class="MuiTable-root-079 UserTable-table-050"
  >
    <thead
      class="MuiTableHead-root-080"
    >
      <tr
        class="MuiTableRow-root-081 MuiTableRow-head-084"
      >
        <th
          class="MuiTableCell-root-088 MuiTableCell-head-089 WithStyles-TableCell--head-086"
          scope="col"
        />
        <th
          class="MuiTableCell-root-088 MuiTableCell-head-089 WithStyles-TableCell--head-086"
          scope="col"
        >
          First
        </th>
        <th
          class="MuiTableCell-root-088 MuiTableCell-head-089 WithStyles-TableCell--head-086"
          scope="col"
        >
          Last
        </th>
        <th
          class="MuiTableCell-root-088 MuiTableCell-head-089 WithStyles-TableCell--head-086"
          scope="col"
        >
          Email
        </th>
        <th
          class="MuiTableCell-root-088 MuiTableCell-head-089 WithStyles-TableCell--head-086"
          scope="col"
        >
          Registration
        </th>
      </tr>
    </thead>
    <tbody
      class="MuiTableBody-root-096"
    />
  </table>
  <div
    class="Component-root-097"
  >
    <svg
      class="mdi-icon "
      width="64"
      height="64"
      fill="currentColor"
      viewBox="0 0 24 24"
    >
      <path
        d="M15.5,12C18,12 20,14 20,16.5C20,17.38 19.75,18.21 19.31,18.9L22.39,22L21,23.39L17.88,20.32C17.19,20.75 16.37,21 15.5,21C13,21 11,19 11,16.5C11,14 13,12 15.5,12M15.5,14C14.12,14 13,15.12 13,16.5C13,17.88 14.12,19 15.5,19C16.88,19 18,17.88 18,16.5C18,15.12 16.88,14 15.5,14M10,4C12.21,4 14,5.79 14,8C14,8.91 13.69,9.75 13.18,10.43C12.32,10.75 11.55,11.26 10.91,11.9L10,12C7.79,12 6,10.21 6,8C6,5.79 7.79,4 10,4M2,20V18C2,15.88 5.31,14.14 9.5,14C9.18,14.78 9,15.62 9,16.5C9,17.79 9.38,19 10,20H2Z"
      />
    </svg>
    <p>
      There are currently no registered users.
    </p>
  </div>
</div>
`;

exports[`UserTable should render with required props 1`] = `
<div
  class="MuiPaper-root-04 MuiPaper-elevation2-08 MuiPaper-rounded-05 UserTable-root-01"
>
  <table
    class="MuiTable-root-031 UserTable-table-02"
  >
    <thead
      class="MuiTableHead-root-032"
    >
      <tr
        class="MuiTableRow-root-033 MuiTableRow-head-036"
      >
        <th
          class="MuiTableCell-root-040 MuiTableCell-head-041 WithStyles-TableCell--head-038"
          scope="col"
        />
        <th
          class="MuiTableCell-root-040 MuiTableCell-head-041 WithStyles-TableCell--head-038"
          scope="col"
        >
          First
        </th>
        <th
          class="MuiTableCell-root-040 MuiTableCell-head-041 WithStyles-TableCell--head-038"
          scope="col"
        >
          Last
        </th>
        <th
          class="MuiTableCell-root-040 MuiTableCell-head-041 WithStyles-TableCell--head-038"
          scope="col"
        >
          Email
        </th>
        <th
          class="MuiTableCell-root-040 MuiTableCell-head-041 WithStyles-TableCell--head-038"
          scope="col"
        >
          Registration
        </th>
      </tr>
    </thead>
    <tbody
      class="MuiTableBody-root-048"
    >
      <tr
        class="MuiTableRow-root-033 UserTable-row-03"
      >
        <th
          class="MuiTableCell-root-040 MuiTableCell-body-042 WithStyles-TableCell--body-039"
          scope="row"
        >
          <div
            class=" sb-avatar"
            style="display: inline-block; vertical-align: middle; width: 34px; height: 34px; border-radius: 100%; font-family: Helvetica, Arial, sans-serif;"
          >
            <img
              class=" sb-avatar__image"
              width="34px"
              height="34px"
              style="max-width: 100%; width: 34px; height: 34px; border-radius: 100%;"
              src="https://d3iw72m71ie81c.cloudfront.net/male-52.jpg"
              alt="Bob Smith"
            />
          </div>
        </th>
        <td
          class="MuiTableCell-root-040 MuiTableCell-body-042 WithStyles-TableCell--body-039"
        >
          Bob
        </td>
        <td
          class="MuiTableCell-root-040 MuiTableCell-body-042 WithStyles-TableCell--body-039"
        >
          Smith
        </td>
        <td
          class="MuiTableCell-root-040 MuiTableCell-body-042 WithStyles-TableCell--body-039"
        >
          bob.smith@gmail.com
        </td>
        <td
          class="MuiTableCell-root-040 MuiTableCell-body-042 WithStyles-TableCell--body-039"
          style="font-size: 0.875rem;"
        >
          2018 Sept 04
        </td>
      </tr>
      <tr
        class="MuiTableRow-root-033 UserTable-row-03"
      >
        <th
          class="MuiTableCell-root-040 MuiTableCell-body-042 WithStyles-TableCell--body-039"
          scope="row"
        >
          <div
            class=" sb-avatar"
            style="display: inline-block; vertical-align: middle; width: 34px; height: 34px; border-radius: 100%; font-family: Helvetica, Arial, sans-serif;"
          >
            <img
              class=" sb-avatar__image"
              width="34px"
              height="34px"
              style="max-width: 100%; width: 34px; height: 34px; border-radius: 100%;"
              src="https://randomuser.me/api/portraits/women/21.jpg"
              alt="Jane Doe"
            />
          </div>
        </th>
        <td
          class="MuiTableCell-root-040 MuiTableCell-body-042 WithStyles-TableCell--body-039"
        >
          Jane
        </td>
        <td
          class="MuiTableCell-root-040 MuiTableCell-body-042 WithStyles-TableCell--body-039"
        >
          Doe
        </td>
        <td
          class="MuiTableCell-root-040 MuiTableCell-body-042 WithStyles-TableCell--body-039"
        >
          jane.doe@gmail.com
        </td>
        <td
          class="MuiTableCell-root-040 MuiTableCell-body-042 WithStyles-TableCell--body-039"
          style="font-size: 0.875rem;"
        >
          2018 Sept 04
        </td>
      </tr>
      <tr
        class="MuiTableRow-root-033 UserTable-row-03"
      >
        <th
          class="MuiTableCell-root-040 MuiTableCell-body-042 WithStyles-TableCell--body-039"
          scope="row"
        >
          <div
            class=" sb-avatar"
            style="display: inline-block; vertical-align: middle; width: 34px; height: 34px; border-radius: 100%; font-family: Helvetica, Arial, sans-serif;"
          >
            <img
              class=" sb-avatar__image"
              width="34px"
              height="34px"
              style="max-width: 100%; width: 34px; height: 34px; border-radius: 100%;"
              src="https://twostoryrobot.com/not-a-valid-url"
              alt="James Bond"
            />
          </div>
        </th>
        <td
          class="MuiTableCell-root-040 MuiTableCell-body-042 WithStyles-TableCell--body-039"
        >
          James
        </td>
        <td
          class="MuiTableCell-root-040 MuiTableCell-body-042 WithStyles-TableCell--body-039"
        >
          Bond
        </td>
        <td
          class="MuiTableCell-root-040 MuiTableCell-body-042 WithStyles-TableCell--body-039"
        >
          james.bond@gmail.com
        </td>
        <td
          class="MuiTableCell-root-040 MuiTableCell-body-042 WithStyles-TableCell--body-039"
          style="font-size: 0.875rem;"
        >
          2018 Sept 04
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
